gma/integration_test/src/integration_test.cc

```diff
@@ -208,6 +232,141 @@ void InitializeGma(firebase::App* shared_app) {
   LogDebug("Successfully initialized GMA.");
 }
 
+void LogAdResultIfError(const firebase::gma::AdResult* ad_result) {
+  if (ad_result->is_successful() == false) {
+    const firebase::gma::AdError& ad_error = ad_result->ad_error();
+    LogDebug(
+        "AdResult Error code: %d  domain: \"%s\" code: \"%s\" message:"
+        " \"%s\" to_string: \"%s\"",
+        ad_error.code(), ad_error.domain().c_str(), ad_error.message().c_str(),
+        ad_error.ToString().c_str());
+  }
+}
+
+// Attempts to load an AdView configured with the provided listeners.
+// If the service returns a NoFill error, then this function will continue to
+// reattempt the load ad until kMaxNoFillRetries is reached. If
+// kMaxNoFillRetries has been reached, either via this invocation or
+// preveiously, then this function returns the AdResult with the loadAd failure.
+firebase::gma::AdView* loadAdView(
+    const firebase::gma::AdSize& ad_size,
+    const firebase::gma::AdRequest& request,
+    firebase::Future<firebase::gma::AdResult>& load_ad_future,
+    firebase::gma::AdListener* ad_listener,
+    firebase::gma::PaidEventListener* paid_event_listener) {
+  firebase::gma::AdView* ad_view = nullptr;
+
+  do {
```

Comment:
This seems like something that would be ripe to use RunFlakyBlock on rather than having the retry logic replicated here.

Something like...
bool ad_load_successful = RunFlakyBlock([&]() {
  load_ad_future =  ad_view->LoadAd(request);
  FirebaseGmaTest::WaitForCompletionAnyResult(load_ad_future, "LoadAd");
  if (load_ad_future.error() == firebase::gma::kAdErrorCodeNoFill || load_ad_future.error() == firebase::gma::kAdErrorCodeInvalidRequest) {
    // do whatever setup you need to retry the LoadAd request
    return false;  // will trigger a retry
  }
  return (load_ad_future.error() == 0);
}, "LoadAd");

This will do a simple exponential backoff retry until the bool returns true or it fails enough times.

Reply:
I'll give this a shot on Wednesday.